### PR TITLE
refactoring, rename, track by, pipe, custom attribute directive

### PR DESCRIPTION
**Pull request by StanislavYurchenko**

### CODE REVIEW

File: apps/todos/src/app/app.module.ts [hunk 1]
```diff
@@ -2,29 +2,31 @@ import { NgModule } from '@angular/core';
...
-import { ServicesModule } from '@learning-workspace/services';
-// import { ApiService } from 'libs/services/src/lib/api-services/api-services.service';
+import { StopPropagationDirective } from './core/directives';
 
 @NgModule({
   declarations: [
     AppComponent,
     AboutComponent,
-    TodoComponent,
+    TodoDetailsComponent,
```
Comment (ZhmenZH) [inline apps/todos/src/app/app.module.ts]: why do we add it here?

File: apps/todos/src/app/features/todo-details/todo-details.component.ts [hunk 2]
```diff
@@ -1,42 +1,50 @@
...
         this.todo = todo;
-        this.loading = true;
+        this.loading = false;
       });
     this.subscription.add(getTodoSubscription);
   }
 
-  ngOnDestroy(): void {
-    this.subscription.unsubscribe();
+  private getTodoIdFromRoute() {
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todo-details/todo-details.component.ts]: add return function type

File: apps/todos/src/app/features/todos/todos.component.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    this.editMode = true;
+    this.selectedTodo = this.todos.find((todo) => todo.id === id) as Todo;
+    this.openAddEditTodoForm();
+  }
+
+  public checkTodo(id: string): void {
+    this.checkTodoSubscription(id);
+  }
+
+  public openAddEditTodoForm() {
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/todos.component.ts]: please add return function type to all functions in this component

File: apps/todos/src/app/features/todos/todos.component.html [hunk 4]
```diff
@@ -0,0 +1,44 @@
+<ul *ngIf="todos?.length" class="list-group">  
+  <ng-container *ngFor="let todo of todos; trackBy:todoIdentify">
+    <li role="button"
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/todos.component.html]: please create separate component for one todo item and use it here

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+import { AddTodo } from '@learning-workspace/api-interfaces';
+
+@Component({
+  selector: 'learning-workspace-add-edit-todo-form',
+  templateUrl: './add-edit-todo-form.component.html',
+  styleUrls: ['./add-edit-todo-form.component.scss'],
+})
+export class AddEditTodoFormComponent implements OnInit, AfterViewInit {
+  public todoForm: FormGroup;
```
Comment (ZhmenZH) [inline apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: please use this property below Input() and Output()

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  constructor(
+    private readonly fb: FormBuilder,
+    private readonly modalService: NgbModal
+  ) {}
+
+  ngOnInit(): void {
+    this.todoForm = this.AddEditTodoFormBuild();
+  }
+
+  ngAfterViewInit() {
```
Comment (ZhmenZH) [inline apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: please add return function type

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    this.openAddTodoForm(this.content);
+  }
+
+  private AddEditTodoFormBuild(): FormGroup {
```
Comment (ZhmenZH) [inline apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: please rename to addEditTodoFormBuild

**Review by ZhmenZH — CHANGES_REQUESTED** [review]

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts (hunk 7 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: it was renamed

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts (hunk 6 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: Did it

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts (hunk 5 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: did it

File: apps/todos/src/app/features/todos/todos.component.html (hunk 4 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/todos.component.html]: I did it

File: apps/todos/src/app/features/todos/todos.component.ts (hunk 3 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/todos.component.ts]: ready

File: apps/todos/src/app/features/todo-details/todo-details.component.ts (hunk 2 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todo-details/todo-details.component.ts]: I did it

File: apps/todos/src/app/core/directives/stop-propagation/stop-propagation.directive.ts [hunk 8]
```diff
@@ -0,0 +1,11 @@
+import { Directive, HostListener } from '@angular/core';
```
Comment (ZhmenZH) [inline apps/todos/src/app/core/directives/stop-propagation/stop-propagation.directive.ts:1]: In the directives directory create module directives and declare stop-propagation directive there, then import directives module where you want to use your directive

File: apps/todos/src/app/app.module.ts [hunk 9]
```diff
@@ -2,27 +2,19 @@ import { NgModule } from '@angular/core';
...
-// import { ApiService } from 'libs/services/src/lib/api-services/api-services.service';
 
 @NgModule({
   declarations: [
     AppComponent,
-    AboutComponent,
-    TodoComponent,
-    TodosComponent,
-    PageNotFoundComponent,
     HeaderComponent,
```
Comment (ZhmenZH) [inline apps/todos/src/app/app.module.ts]: Maybe it is better to create a module for the footer and header?

File: apps/todos/src/app/features/todo-details/containers/todo-details/todo-details.component.ts [hunk 10]
```diff
@@ -1,42 +1,49 @@
...
 import { Todo } from '@learning-workspace/api-interfaces';
+import { ApiService } from '../../../../core/services';
 
 @Component({
-  selector: 'learning-workspace-todo',
-  templateUrl: './todo.component.html',
-  styleUrls: ['./todo.component.scss'],
+  selector: 'learning-workspace-todo-details',
+  templateUrl: './todo-details.component.html',
+  styleUrls: ['./todo-details.component.scss'],
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todo-details/containers/todo-details/todo-details.component.ts:11]: do we need scss file?

File: apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
+import { Component, EventEmitter, Input, Output } from '@angular/core';
+
+import { Todo } from '@learning-workspace/api-interfaces';
+import { TodoActionEnum } from '../../enums';
+
+@Component({
+  selector: 'learning-workspace-todo-item',
+  templateUrl: './todo-item.component.html',
+  styleUrls: ['./todo-item.component.scss'],
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts:9]: do we need scss file?

File: apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+})
+export class TodoItemComponent {
+  @Input() todo: Partial<Todo>;
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts:13]: please change the Change Detection Strategy to onPush

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+import { Subscription } from 'rxjs';
+
+import { Todo, AddTodo } from '@learning-workspace/api-interfaces';
+import { ApiService } from '../../../../core/services';
+import { TodoActionEnum } from '../../enums';
+
+@Component({
+  selector: 'learning-workspace-todos-container',
+  templateUrl: './todos-container.component.html',
+  styleUrls: ['./todos-container.component.scss'],
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: do we need the scss file?

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+    this.openAddEditTodoForm();
+  }
+
+  private checkTodo(id: string): void {
+    this.checkTodoSubscription(id);
+  }
+
+  // TODO: change subscription to rxjs
+  private getTodosSubscription(): void {
+    const getTodosSubscription = this.apiService
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: why do we need this const instead we can do like this this.subscription.add(your code)

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+      .getTodos()
+      .subscribe((todos) => (this.todos = todos));
+    this.subscription.add(getTodosSubscription);
+  }
+
+  private addTodoSubscription(todo: AddTodo): void {
+    const getTodosSubscription = this.apiService
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: the same as above

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+      .addTodo(todo)
+      .subscribe((todo) => this.todos.push(todo));
+    this.subscription.add(getTodosSubscription);
+  }
+
+  private editTodoSubscription(id: string, todo: AddTodo): void {
+    const removeTodosSubscription = this.apiService
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: the same as above

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+      .subscribe((updatedTodo) => {
+        this.todos = this.todos.map((todo) =>
+          todo.id === updatedTodo.id ? updatedTodo : todo
+        );
+      });
+    this.subscription.add(removeTodosSubscription);
+  }
+
+  private checkTodoSubscription(id: string): void {
+    const checkTodosSubscription = this.apiService
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: the same as above

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+      .subscribe((updatedTodo) => {
+        this.todos = this.todos.map((todo) =>
+          todo.id === updatedTodo.id ? updatedTodo : todo
+        );
+      });
+    this.subscription.add(checkTodosSubscription);
+  }
+
+  private removeTodoSubscription(id: string): void {
+    const removeTodosSubscription = this.apiService
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: the same as above

File: apps/todos/src/app/features/todos/todos.module.ts [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+import { TodosRouterModule } from './todos-routing.module';
+
+import { TodosContainerComponent } from './containers';
+import { TodoItemComponent } from './components';
+
+@NgModule({
+  declarations: [
+    TodosContainerComponent,
+    TodoItemComponent,
+    StopPropagationDirective,
```
Comment (ZhmenZH) [inline apps/todos/src/app/features/todos/todos.module.ts]: import it like module

File: apps/todos/src/app/modules/about/containers/about/about.component.ts [hunk 20]
```diff
@@ -0,0 +1,8 @@
+import { Component } from '@angular/core';
+
+@Component({
+  selector: 'learning-workspace-about',
+  templateUrl: './about.component.html',
+  styleUrls: ['./about.component.scss'],
+})
+export class AboutComponent { }
```
Comment (ZhmenZH) [inline apps/todos/src/app/modules/about/containers/about/about.component.ts]: do we need about component?

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts [hunk 21]
```diff
@@ -0,0 +1,71 @@
+import { Component, Input, OnInit, AfterViewInit, TemplateRef, ViewChild, Output, EventEmitter } from '@angular/core';
+import { FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
+
+import { AddTodo } from '@learning-workspace/api-interfaces';
+
+@Component({
+  selector: 'learning-workspace-add-edit-todo-form',
+  templateUrl: './add-edit-todo-form.component.html',
+  styleUrls: ['./add-edit-todo-form.component.scss'],
```
Comment (ZhmenZH) [inline apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: make this component OnPush strategy

**Review by ZhmenZH — CHANGES_REQUESTED** [review]

File: apps/todos/src/app/modules/about/containers/about/about.component.ts (hunk 20 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/modules/about/containers/about/about.component.ts]: Yes it is, finally here will be information about this project

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts (hunk 13 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: I believe yes, I plan to add styles finally

File: apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts (hunk 11 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts:9]: I believe yes, I plan to add styles finally

File: apps/todos/src/app/features/todo-details/containers/todo-details/todo-details.component.ts (hunk 10 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todo-details/containers/todo-details/todo-details.component.ts:11]: I believe yes, I plan to add styles finally

File: apps/todos/src/app/app.module.ts (hunk 1 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/app.module.ts]: I change it

File: apps/todos/src/app/core/directives/stop-propagation/stop-propagation.directive.ts (hunk 8 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/core/directives/stop-propagation/stop-propagation.directive.ts:1]: Thanks, I did it

File: apps/todos/src/app/app.module.ts (hunk 9 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/app.module.ts]: Yes, I did it

File: apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts (hunk 12 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/components/todo-item/todo-item.component.ts:13]: Thanks, I did it

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts (hunk 14 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: refactored it

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts (hunk 15 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: refactored it

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts (hunk 16 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: refactored it

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts (hunk 17 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: refactored it

File: apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts (hunk 18 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/containers/todos-container/todos-container.component.ts]: refactored it

File: apps/todos/src/app/features/todos/todos.module.ts (hunk 19 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/features/todos/todos.module.ts]: Yes, I did it

File: apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts (hunk 21 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/modules/add-edit-todo-form/add-edit-todo-form.component.ts]: Yes I did it

File: apps/todos/src/app/core/directives/directives.module.ts [hunk 22]
```diff
@@ -0,0 +1,15 @@
...
+import { CommonModule } from '@angular/common';
+
+import { StopPropagationDirective } from './stop-propagation';
+
+const directives = [
+  StopPropagationDirective,
+]
+
+@NgModule({
+  declarations: directives,
```
Comment (ZhmenZH) [inline apps/todos/src/app/core/directives/directives.module.ts:11]: You don't need to do like this here, change it please)

**Review by ZhmenZH — CHANGES_REQUESTED** [review]

File: apps/todos/src/app/core/directives/index.ts [hunk 23]
```diff
@@ -0,0 +1 @@
+export * from './directives.module';
```
Comment (ZhmenZH) [inline apps/todos/src/app/core/directives/index.ts:1]: you don't need to add an index file only for the module, from my side I really don't see an advantage for this case (with modules), please remove everywhere from application in the same cases

**Review by ZhmenZH — CHANGES_REQUESTED** [review]

File: apps/todos/src/app/core/directives/index.ts (hunk 23 above)
Comment (StanislavYurchenko) [reply apps/todos/src/app/core/directives/index.ts:1]: Thanks, I did it

**Pull request merged**